qualtran/bloqs/qsp/fast_qsp_test.py

```diff
@@ -0,0 +1,127 @@
+            P, Q, random_state=random_state, rtol=precision
+        )
+
+
+@pytest.mark.parametrize("degree, precision", [(3, 1e-4), (4, 1e-4)])
+def test_real_polynomial_has_real_complementary_polynomial_quick(degree: int, precision: float):
+    random_state = np.random.RandomState(42)
+
+    for _ in range(10):
+        P = random_qsp_polynomial(degree, random_state=random_state, only_real_coeffs=True)
+        Q = fast_complementary_polynomial(P, only_reals=True)
+        Q = np.around(Q, decimals=8)
+        assert np.isreal(Q).all()
+        check_polynomial_pair_on_random_points_on_unit_circle(
+            P, Q, random_state=random_state, rtol=precision
+        )
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize(
+    "degree, num_tests, precision", [(5, 20, 2e-5), (10, 20, 2e-5), (20, 5, 2e-5), (30, 3, 2e-5)]
+)
+def test_complementary_polynomial(degree: int, num_tests: int, precision: float):
+    random_state = np.random.RandomState(42)
+
+    results = []
+    for _ in range(num_tests):
+        P = random_qsp_polynomial(degree, random_state=random_state)
+        Q = fast_complementary_polynomial(P)
+        results.append(np.array(Q))
+        check_polynomial_pair_on_random_points_on_unit_circle(
+            P, Q, random_state=random_state, rtol=precision
+        )
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize("degree, num_tests", [(2, 10), (3, 10), (4, 10), (5, 10), (10, 10)])
+def test_fast_real_qsp_with_symbolic_signal_matrix(degree: int, num_tests: int):
+    random_state = np.random.RandomState(102)
+
+    for _ in range(num_tests):
+        P = random_qsp_polynomial(degree, random_state=random_state)
+        U = MatrixGate.random(2, random_state=random_state)
+        Q = fast_complementary_polynomial(P)
+        verify_generalized_qsp(U, P, Q=Q)
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize(
+    "degree, num_tests, precision",
+    [(2, 10, 2e-2), (3, 10, 2e-2), (4, 10, 2e-2), (5, 10, 2e-2), (10, 10, 2e-2), (20, 10, 2e-2)],
+)
+def test_real_polynomial_has_real_complementary_polynomial(
+    degree: int, num_tests: int, precision: float
+):
+    random_state = np.random.RandomState(42)
+    for _ in range(num_tests):
+        P = random_qsp_polynomial(degree, random_state=random_state, only_real_coeffs=True)
+        Q = fast_complementary_polynomial(P, only_reals=True)
+        Q = np.around(Q, decimals=8)
+        assert np.isreal(Q).all()
+        check_polynomial_pair_on_random_points_on_unit_circle(
+            P, Q, random_state=random_state, rtol=precision
+        )
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize(
+    "bitsize, degree, negative_power, tolerance",
+    [
+        (1, 2, 0, 1e-5),
+        (1, 2, 1, 1e-5),
+        (1, 2, 2, 1e-5),
+        (2, 2, 0, 1e-5),
+        (2, 2, 1, 1e-5),
+        (2, 2, 2, 1e-5),
+        (1, 5, 2, 1e-4),
+        (1, 5, 0, 1e-4),
+        (3, 5, 0, 1e-4),
+        (3, 5, 2, 1e-4),
+        (2, 20, 1, 1e-1),
+    ],
+)
```

Comment:
I think you can parametrize `bitsize` separately because the complexity of computing Q and verifying P,Q only add up (with bitsize affecting only the latter). You can just use `@pytest.mark.parametrize("bitsize", [1, 2, 3])`